### DISCUSSION TITLE
Collecting page data  .[ReferenceError: window is not defined] , in nextjs15

### DISCUSSION THREAD

**Discussion by ahmerarain** [vercel/next.js#74428] [App Router] [answered]
i have add all my component  'use client' where I am using window, but still getting problems during build,
i also added  below condition but still getting the same problem 
  if (typeof window == "undefined" || typeof document == "undefined") return;

<img width="491" alt="Screenshot 2024-12-31 at 12 52 18 AM" src="https://github.com/user-attachments/assets/49ddbb02-e4aa-42cd-8fc9-66cf324940ed" />

**Comment by icyJoseph** [comment]
> i have add all my component 'use client' where I am using window, but still getting problems during build,
> i also added below condition but still getting the same problem

`use client` marks a module as a client component, but that doesn't mean it will only execute in the browser:

- https://github.com/reactwg/server-components/discussions/4

Client Components are still SSR'd.

Could you show us more of how the code is accessing `window`?

**Reply by ahmerarain** [reply]
<img width="542" alt="Screenshot 2024-12-31 at 1 59 07 AM" src="https://github.com/user-attachments/assets/7191dd71-46be-4578-9ea9-9964c537b6d6" />
<img width="416" alt="Screenshot 2024-12-31 at 1 59 26 AM" src="https://github.com/user-attachments/assets/ec145ffb-0cbd-4590-aee2-70ef529ecb93" />
<img width="701" alt="Screenshot 2024-12-31 at 2 00 35 AM" src="https://github.com/user-attachments/assets/b6eb7ce2-456e-4079-b755-b6e00eba1690" />
<img width="652" alt="Screenshot 2024-12-31 at 2 01 27 AM" src="https://github.com/user-attachments/assets/32170a8d-efed-419a-a504-c1ff571e1cd9" />

**Reply by icyJoseph** [reply]
🤔 I guess one of those imports is using `window` then..., but which one... and how... got a link to this repository?

**Reply by ahmerarain** [reply]
yes, you are right, but on every component that contains window, I added there 'use client'   and condition to check window type is not undefined

**Reply by icyJoseph** [reply]
Yeah, but `use client` doesn't skip SSR. 

The `typeof` should have worked, but I'd like to see how it is being used. Clearly you still have - some unguarded access. Are you always using `typeof`?

**Reply by ahmerarain** [reply]
yes,  always using typeof, like below screenshot 

<img width="747" alt="Screenshot 2024-12-31 at 11 24 42 AM" src="https://github.com/user-attachments/assets/b97a9f26-6266-44a5-84f1-959cc4257786" />


but i  will check again and then get back to you  , thx :)

**Reply by icyJoseph** [reply, marked as answer]
You don't need those checks in the screenshot. The window object, and document, will be available within the useEffect. Again, only within the useEffect, because it runs only on the browser. Same with local storage, but once again, only inside the useEffect.

Maybe a library you use, used window without guards?

**Reply by icyJoseph** [reply]
Also make sure you don't access the window object, at the top level scope of a module, that you then import. Importing a module, runs the top level scope code, and if the window object is referenced there, it'll cause issues too.

**Reply by ahmerarain** [reply]
that makes sense, 
 i will check and refactor the code 
 
 Thx for the help joseph

**Reply by ahmerarain** [reply]
thx, joseph its worked, now 
i made a separate hook for window and document ,


`"use client";
import { useState, useEffect } from "react";

const useClientSide = () => {
  const [clientWindow, setWindow] = useState(undefined);
  const [clientDocument, setDocument] = useState(undefined);

  useEffect(() => {
    setWindow(window);
    setDocument(document);
  }, []);

  return { window: clientWindow, document: clientDocument };
};

export default useClientSide;
`
